### PR TITLE
Fixes frozen water being considered in washing clothes.

### LINKED ISSUE #26234: Cleaning using a washboard ignores water frozen status
**Describe the bug**  
The user is able to clean clothes etc using a washboard when the only available source of water is in a frozen status. If you can't drink it, you surely can't clean with it. Someone missed adding a status check somewhere.

**To Reproduce**  
1. Stand by a toilet with frozen water in it and the right supplies to clean
2. Use a washboard to clean 
3. Observe that the cleaning operation was successful 

**Expected behavior**  
1. The cleaning operation should fail due to the frozen water being incompatible with cleaning

**Screenshots**  

**Versions and configuration(please complete the following information):**
Running build 8046 on Windows 7

### PR DESCRIPTION
**Pull request by metiscus**
Fixed issue where frozen water was being counted towards clothing being washed

#### Summary
```SUMMARY: Bugfixes "Prevents frozen water from being considered in washing clothes"```

#### Purpose of change
Resolves #26234
#### Describe the solution
Adds a filter function to inventory charge search.
Hooks up the action of washing to utilize new inventory filtered search. 

#### Additional context
The consume charges function will still consume frozen water at this time. I plan on fixing that as well.
This is my first attempt at contributing and I know that this is probably not the correct way to implement what was needed so I am open to any assistance / comments you may have.

### CODE REVIEW

**Comment by metiscus** [comment]
I suspect this PR won't stop the use of groundwater that is frozen. Is there a good way to whip up a test case for this? The map editor doesn't seem to let me modify the terrain to be a frozen pond.

File: src/inventory.cpp [hunk 1]
```diff
@@ -19,6 +19,23 @@
 #include "output.h"
 #include "translations.h"
 
+/** Sums the two terms, being careful to not trigger overflow.
+ * Doesn't handle underflow.
```
[the remark this answers is not in the document]
Comment (metiscus) [reply src/inventory.cpp]: This function was copied directly from the internals of the inventory class. I'll link the exact location but perhaps we should address both locations and optionally deduplicate the code.

File: src/iuse.cpp [hunk 2]
```diff
@@ -7671,7 +7671,7 @@ int iuse::washclothes( player *p, item *it, bool, const tripoint & )
     const inventory &crafting_inv = p->crafting_inventory();
     if( !crafting_inv.has_charges( "water", required_water ) &&
         !crafting_inv.has_charges( "water_clean", required_water ) ) {
```
[the remark this answers is not in the document]
Comment (metiscus) [reply src/iuse.cpp]: Agreed. I'll address it.

File: src/inventory.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (jbytheway) [reply src/inventory.cpp]: Maybe belongs in `cata_utility.h`.

**Comment by dissociativity** [comment]
perhaps since we don't wash clothes super frequently, and it'd be more convenient, I'd suggest adding a check for a nearby fire + boiling item or mess kit, ect?

File: src/cata_utility.h [hunk 3]
```diff
@@ -72,6 +74,31 @@ double round_up( double val, unsigned int dp );
  */
 bool isBetween( int test, int down, int up );
 
+/** Sums the two terms, being careful to not trigger overflow or underflow
+ * Correctly handles overflow, underflow, and nonintegral types
+ *
+ * @param a The first addend.
```
Comment (BevapDin) [inline src/cata_utility.h]: Is a separate explicit documentation of `a` and `b` really necessary? Isn't it clearly documented and obvious what those two values are via the main function documentation above?

File: src/cata_utility.h [hunk 4]
```diff
@@ -72,6 +74,31 @@ double round_up( double val, unsigned int dp );
...
+ * @param b The second addend.
+ * @return the sum of the addends, but truncated to std::numeric_limits<T>::max or min.
+ */
+template <typename T>
+static T sum_no_wrap( T a, T b )
+{
+    static_assert( std::is_integral<T>::value, "sum_no_wrap applied to nonintegral type." );
+
+    // Check for overflow
+    if( ( a > 0 ) && ( b > std::numeric_limits<T>::max() - a ) ) {
```
Comment (BevapDin) [inline src/cata_utility.h]: The parenthesis around the relational operators are not needed:
```suggestion
    if( a > 0 && b > std::numeric_limits<T>::max() - a ) {
```

File: src/inventory.cpp [hunk 5]
```diff
@@ -690,6 +691,17 @@ bool inventory::has_charges( const itype_id &it, long quantity ) const
...
 
+bool inventory::has_charges_with( const itype_id &it, long quantity,
+                                  const std::function<bool( const item & )> &filter ) const
+{
+    long charges = 0;
+    for( const item *itr : items_with( filter ) ) {
+        charges = sum_no_wrap( charges, itr->charges_of( it ) );
+    }
+
+    return ( charges >= quantity );
```
Comment (BevapDin) [inline src/inventory.cpp]:
```suggestion
    return charges >= quantity;
```

File: src/visitable.cpp [hunk 6]
```diff
@@ -12,6 +12,10 @@
 #include "string_id.h"
 #include "submap.h"
 #include "veh_type.h"
+#include "game.h"
```
Comment (BevapDin) [inline src/visitable.cpp]: The include of "cata_utility.h", I buy. But why all the other additional includes?

File: src/inventory.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (metiscus) [reply src/inventory.cpp]: I addressed your first comment and moved the common implementation where you suggested.

File: src/visitable.cpp (hunk 6 above)
Comment (metiscus) [reply src/visitable.cpp]: I think I got something wrong with my merge... I only intentionally added cata_utility.h

**Pull request closed without merging**